feat(tui): Allow keybinding of custom slash commands (resolves #5904)

ariane-emory opened the pull request:
### What does this PR do?

Adds support for setting key bindings for custom slash commands. Allows users to bind custom slash commands to keystrokes by adding
entries to the keybinds config with keys starting with '/'. When one of these keys is struck, command is inserted before the text in the main prompt text input box and then that input is submitted, allowing arguments to be passed to custom slash commands as normal (e.e., if the text input box contains `foo "bar"` and a key is struck to trigger `/my-command`, then `/my-command foo "bar"` is submitted`).

If the slash command referenced by a keybinding does not exist when they key is struck (e.g., if the command's `.md` file was deleted while opencode was running) this is handled safely, and a toast is displayed indicating that no such command existsts.

I've been using this feature all week in my personal fork and it has made a truly _tremendous_ difference in my workflow, i really think that this one is a killer feature.

### Changes:
- Update Config.Keybinds schema to accept arbitrary keys using catchall
- Add keyboard handler in TUI to match and execute command keybinds
- Regenerate TypeScript SDK to reflect updated KeybindsConfig type

When a command keybind is pressed, the prompt is set to the command
text and immediately submitted, executing the custom command.

### Example usage in config.json:

```json
{
  "$schema": "https://opencode.ai/config.json",
  "keybinds": {
    "/commit-and-push--using-big-pickle": "ctrl+alt+c",
    "/fetch-and-pull-all-branches--using-glm-4.6": "ctrl+alt+f",
    "/merge-dev-into-branch-resolve-conflicts-and-push--using-copilot-sonnet": "ctrl+alt+m",
    "/integrate-the-branches-listed-in-yoinklistmd--using-antigravity-opus": "ctrl+alt+i",
    "/integrate-the-branches-listed-in-yoinklistmd--using-copilot-opus": "ctrl+alt+shift+i",
  }
}
```

Resolves #5904.

ariane-emory committed:
feat: add keybind support for custom slash commands

Allow users to bind custom slash commands to keystrokes by adding
entries to the keybinds config with keys starting with '/'.

Changes:
- Update Config.Keybinds schema to accept arbitrary keys using catchall
- Add keyboard handler in TUI to match and execute command keybinds
- Regenerate TypeScript SDK to reflect updated KeybindsConfig type

When a command keybind is pressed, the prompt is set to the command
text and immediately submitted, executing the custom command.

Example usage in config.json:
{
  "keybinds": {
    "/my-command": "ctrl+shift+m"
  }
}

Resolves: #47

ariane-emory committed:
fix: handle undefined values in command keybind parsing

Add null checks to prevent TypeScript errors when keybind values
are undefined in the config.

ariane-emory committed:
feat: preserve prompt text as command arguments in keybind execution

When a command keybind is triggered, preserve any existing text in the
prompt input box and append it as arguments to the command.

Example: If the prompt contains 'blah blah abrahadabra' and the user
presses a key bound to /foo, the prompt becomes '/foo blah blah abrahadabra'
before submission, passing the original text as arguments to the command.

This makes command keybinds more flexible and allows users to quickly
apply commands to text they've already typed.

ariane-emory committed:
Merge dev branch, resolving types.gen.ts conflict

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
fix: remove unused sdk variable in app.tsx

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/revise-keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'feat/keybindable-commands' of github.com:ariane-emory/opencode into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'feat/keybindable-commands' of github.com:ariane-emory/opencode into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Fix type compatibility for vercel provider after merge

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
revert a file

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Fix type compatibility with Vercel AI SDK provider

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
revert a file

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

rekram1-node committed:
fix: ensure 'name' isnt being sent in request body for custom agent

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

rekram1-node committed:
test: fix test

ariane-emory committed:
Merge remote-tracking branch 'origin/dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'feat/keybindable-commands' of github.com:ariane-emory/opencode into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
fix: update permission config test to match new key order from dev merge

ariane-emory committed:
Merge branch 'feat/keybindable-commands' of github.com:ariane-emory/opencode into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
fix: revert file

ariane-emory committed:
Merge remote-tracking branch 'origin/dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge dev into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge dev into feat/keybindable-commands

ariane-emory committed:
Merge dev into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge dev into feat/keybindable-commands

ariane-emory committed:
Merge dev into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
fix: allow custom keybinds in tui.json by using catchall instead of strict

johnnymo87 reviewed (commented):
Would love to see this merged -- I want to bind `/skills` and `/mcps` to keybinds, and this is the cleanest path to that.

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge dev into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

# Conflicts:
#	packages/opencode/src/cli/cmd/tui/app.tsx

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
fix sdk generation noise for keybind commands

ariane-emory committed:
Merge dev into feat/keybindable-commands

ariane-emory committed:
fix: import Keybind from @/util instead of @/util/keybind to use namespace export

ariane-emory committed:
fix: allow custom command keybinds in keybinds schema

The Effect Schema migration lost the .catchall(z.string()) that allowed
arbitrary string keys like /note-dir for custom command keybinds. Change
Schema.Struct to Schema.StructWithRest with a Record catchall to restore
custom command keybind support.

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
Merge branch 'dev' into feat/keybindable-commands

ariane-emory committed:
merge(dev): merge dev into feat/keybindable-commands

ariane-emory committed:
chore(tui): remove stale openrouter warning dialog

ariane-emory committed:
merge(dev): merge dev into feat/keybindable-commands
